src/sentry_core.c

```diff
@@ -692,3 +735,71 @@ sentry_set_level(sentry_level_t level)
         scope->level = level;
     }
 }
+
+sentry_value_t
+sentry_transaction_start(sentry_value_t tx_cxt)
+{
+    sentry_value_t tx = sentry_value_new_event();
+
+    // TODO: stuff transaction into the scope
+    bool should_sample = sentry__should_send_transaction(tx_cxt);
+    sentry_value_set_by_key(
+        tx, "sampled", sentry_value_new_bool(should_sample));
+
+    sentry_value_set_by_key(
+        tx, "transaction", sentry_value_get_by_key_owned(tx_cxt, "name"));
+    sentry_value_set_by_key(tx, "start_timestamp",
+        sentry__value_new_string_owned(
+            sentry__msec_time_to_iso8601(sentry__msec_time())));
+
+    sentry_uuid_t trace_id = sentry_uuid_new_v4();
+    sentry_value_set_by_key(
+        tx, "trace_id", sentry__value_new_internal_uuid(&trace_id));
+
+    sentry_uuid_t span_id = sentry_uuid_new_v4();
+    sentry_value_set_by_key(
+        tx, "span_id", sentry__value_new_span_uuid(&span_id));
+
+    sentry_value_decref(tx_cxt);
+
+    return tx;
+}
+
+void
+sentry_transaction_finish(sentry_value_t tx)
+{
+    sentry_value_t sampled = sentry_value_get_by_key(tx, "sampled");
+    if (!sentry_value_is_null(sampled) && !sentry_value_is_true(sampled)) {
+        sentry_value_decref(sampled);
```

Comment:
`sentry_value_get_by_key` returns a borrowed ref, vs `sentry_value_get_by_key_owned`. So you don’t need to decref here.
I think this hides a latent bug, since `sampled` is most likely a bool, which is treated as a primitive type and is thus not refcounted at all.

Reply:
iiuc you can safey incref/decref primitive types so that wrt reference counting you can treat all sentry_value_t's the same?

Reply:
seconding @flub here: `sentry_value_decref` appears to "handle" primitives correctly so one could treat all `sentry_value_t`'s the same way.

while we know that this particular instance is likely to be a bool, would it be better to do this here as opposed to omitting it? by ignoring the implementation details it makes the API and its usage consistent across the codebase.

Reply:
I can construct an example that exhibits a double-free:

```
sentry_value_t tx = sentry_value_new_object();
sentry_value_set_by_key(tx, "sampled", sentry_value_new_double(0.0));
```

Yes, this is a very contrived example. But doubles are refcounted (not primitives), and `0.0` is considered `false` and run into this branch.

Reply:
wait, while what i said is correct.  i misunderstood what was going on and now realise the right thing here is still do handle the reference counting correct because we don't know what type `sampled` actually is.  And because `sentry_value_get_by_key` returns a borrowed value we should not be decrefing it.